docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+- **Bootstrapping**: turning a server into a Kubernetes node
+
+## Summary
+
+This proposal outlines splitting up the Machine data model into two new providers/controllers along with the generic Machine controller. The two new providers are the Infrastructure provider and the Bootstrap provider. The infrastructure provider handles provisioning infrastructure (cloud, bare metal, virtual) and the bootstrap provider handles Kubernetes bootstrapping. This change will improve separation of concerns for the Machine controller and gives more degrees of freedom to the user.
+
+In Cluster API v1alpha1, users can create Machine custom resources. Cluster API components react to a new Machine by provisioning infrastructure (typically a physical or virtual machine) and then bootstrapping Kubernetes on the provisioned server, making it a Kubernetes Node. Cluster API’s current architecture mandates that a single “provider” handle both the infrastructure provisioning and the Kubernetes bootstrapping.
+
+## Motivation
+- A singular provider that combines infrastructure provisioning and Kubernetes bootstrapping is too restrictive. Many current v1alpha1 providers target single cloud providers. If a user wants to install a specific Kubernetes distribution (Kubernetes, OpenShift, Rancher, etc.), it is left to the provider to add support for it. In practical terms, we are not able to guarantee that a user can install the same distribution regardless of infrastructure environment.
+- Both users and developers are confused by the code organization split where all the controllers live in the cluster-api repository, but some controllers are part of the generic cluster-api pod (MachineDeployment, MachineSet), while other controllers are part of each provider’s pod (Cluster, Machine). We would like for the code location to correspond to the pod, which hopefully will reduce both user and developer confusion.
+- The Machine custom resource contains fields for provider-specific configuration and status. These are represented as opaque [RawExtensions](https://godoc.org/k8s.io/apimachinery/pkg/runtime#RawExtension), meaning their contents are not directly processed as part of the Machine’s validation. This leads to a poor user experience
+
+### Goals
+1. To use Kubernetes Controllers to manage the lifecycle of Machines.
+1. To make bootstrapping reusable across multiple infrastructure providers.
+1. To support pre-boot (cloud-init, or similar) bootstrapping.
+1. To support bootstrapping different Kubernetes distributions on the same infrastructure provider.
+1. To validate provider-specific content as early as possible (create, update).
+
+### Non-Goals
+1. To modify the Cluster object.
+1. To fully implement machine state lifecycle hooks. This must be part of a future proposal that builds on these proposed changes.
+1. To setup OS configuration, install software or any other features related to image building.
```

Comment:
However, the example we use for boostrap configuration is cloud-init which allows setting up OS configuration and installing required software, therefore making this distinction arbitrary.

Reply:
True, but cloud-init is not a universal standard (e.g. we're considering building an [ignition](https://coreos.com/ignition/)-driven bootstrap provider). I read that statement as saying it's not the Machine's responsibility to define what software need be installed, but instead that belongs to the interaction between the bootstrap provider and the thing that ultimately provisions the resources.

Reply:
Maybe there's a misunderstanding here about who uses the bootstrap information. It is provided by the bootstrap controller and used by the provider specific infrastructure provisioner. Both of these elements live outside CAPI. Therefore, I don't see the point about been opinionated with respect of what the bootstrap data contains or is used.